csmt/csmt.go

```diff
@@ -90,22 +99,144 @@ func (tree *CSMT) doInsert(node Node, key *Key, leafHash *Hash) (Node, error) {
 	}
 
 	if leftDistance < rightDistance {
-		newNode, err := tree.doInsert(left, key, leafHash)
+		newNode, err := tree.doInsert(left, key, leafHash, value)
 		if err != nil {
 			return nil, err
 		}
 		return tree.createInnerNode(newNode, right), nil
 	}
 
-	newNode, err := tree.doInsert(right, key, leafHash)
+	newNode, err := tree.doInsert(right, key, leafHash, value)
 	if err != nil {
 		return nil, err
 	}
 	return tree.createInnerNode(left, newNode), nil
 }
 
-// GetProof gets the proof for key
-func (tree *CSMT) GetProof(key *Key) Proof {
+// GetValue gets the value at leafHash
+func (tree *CSMT) GetValue(leafHash *Hash) (interface{}, error) {
```

Comment:
instead of using interface{} so you can return nil, return a default key value on an error or return a pointer to a key

Reply:
What do you mean by "key" here? I think the key is just the leafHash. The returned value is the data bound to the hash (key). I really can't think the usage of the value. I think we just need to store hashes (keys) in the tree, and verify if they exist.